File: live/src/encoding.ts
/**
 * The browser half of the pose vector encoding.
 *
 * This is a port of `preprocessing/bodypose/encoding.py` and has to stay
 * identical to it — a query vector encoded differently from the index is not a
 * worse match, it is a meaningless one. The parameters that could drift
 * (origin, scale, confidence floor) are not duplicated here as constants; they
 * arrive from the index's own `index.json` and are applied at runtime, and
 * `assertCompatible` refuses an index this code was not written against.
 */

export const ENCODING_VERSION = 'coco17-v1'

export const COCO17 = [
  'nose',
  'left_eye',
  'right_eye',
  'left_ear',
  'right_ear',
  'left_shoulder',
  'right_shoulder',
  'left_elbow',
  'right_elbow',
  'left_wrist',
  'right_wrist',
  'left_hip',
  'right_hip',
  'left_knee',
  'right_knee',
  'left_ankle',
  'right_ankle',
] as const

export const NUM_JOINTS = 17
export const DIMS = NUM_JOINTS * 2
export const MIN_EXTENT = 1e-4

/** Index permutation that swaps left and right. */
export const MIRROR_PERM = [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

/** BlazePose's 33 landmarks contain COCO's 17; this is the gather. */
export const BLAZEPOSE_TO_COCO17 = [
  0, 2, 5, 7, 8, 11, 12, 13, 14, 15, 16, 23, 24, 25, 26, 27, 28,
]

/** Bones, as index pairs into COCO17 — for the skeleton overlay. */
export const SKELETON: [number, number][] = [
  [5, 6], [5, 7], [7, 9], [6, 8], [8, 10],
  [5, 11], [6, 12], [11, 12],
  [11, 13], [13, 15], [12, 14], [14, 16],
  [0, 1], [0, 2], [1, 3], [2, 4], [0, 5], [0, 6],
]

export interface EncodingParams {
  version: string
  origin: 'center' | 'corner'
  scale: 'uniform' | 'independent'
  conf_floor: number
}

export interface EncodedPose {
  /** 34 floats, unit L2 norm. */
  vector: Float32Array
  /** 17 confidences in [0, 1]; zero where the joint was below the floor. */
  weights: Float32Array
  bbox: [number, number, number, number]
  visible: number
}

/**
 * Keypoints are a flat `[x, y, confidence] * 17` array in image-normalised,
 * top-left-origin coordinates — the same convention the indexer uses.
 */
export type Keypoints = Float32Array

export function assertCompatible(params: EncodingParams): void {
  if (params.version !== ENCODING_VERSION) {
    throw new Error(
      `Index was built with encoding "${params.version}" but this build speaks ` +
        `"${ENCODING_VERSION}". Rebuild the index, or check out the matching commit.`,
    )
  }
  if (params.origin !== 'center' && params.origin !== 'corner') {
    throw new Error(`Unknown origin mode "${params.origin}" in index.json`)
  }
  if (params.scale !== 'uniform' && params.scale !== 'independent') {
    throw new Error(`Unknown scale mode "${params.scale}" in index.json`)
  }
}

/**
 * Translate, scale and normalise a pose into the comparable unit vector.
 * Returns null when there is nothing to encode — no confident joints, or a
 * bounding box with no extent in either axis.
 */
export function encode(kp: Keypoints, params: EncodingParams): EncodedPose | null {
  let visible = 0
  let x0 = Infinity
  let y0 = Infinity
  let x1 = -Infinity
  let y1 = -Infinity

  for (let j = 0; j < NUM_JOINTS; j++) {
    const c = kp[j * 3 + 2]
    if (c < params.conf_floor) continue
    visible++
    const x = kp[j * 3]
    const y = kp[j * 3 + 1]
    if (x < x0) x0 = x
    if (x > x1) x1 = x
    if (y < y0) y0 = y
    if (y > y1) y1 = y
  }
  if (visible === 0) return null

  const w = x1 - x0
  const h = y1 - y0
  if (w < MIN_EXTENT && h < MIN_EXTENT) return null

  const ox = params.origin === 'center' ? (x0 + x1) * 0.5 : x0
  const oy = params.origin === 'center' ? (y0 + y1) * 0.5 : y0

  let sx: number
  let sy: number
  if (params.scale === 'uniform') {
    sx = sy = Math.max(w, h, MIN_EXTENT)
  } else {
    sx = Math.max(w, MIN_EXTENT)
    sy = Math.max(h, MIN_EXTENT)
  }

  const vector = new Float32Array(DIMS)
  const weights = new Float32Array(NUM_JOINTS)
  for (let j = 0; j < NUM_JOINTS; j++) {
    const c = kp[j * 3 + 2]
    if (c < params.conf_floor) continue
    // Joints below the floor stay at the origin, which is neutral once weighted.
    vector[j * 2] = (kp[j * 3] - ox) / sx
    vector[j * 2 + 1] = (kp[j * 3 + 1] - oy) / sy
    weights[j] = Math.min(1, Math.max(0, c))
  }

  let norm = 0
  for (let i = 0; i < DIMS; i++) norm += vector[i] * vector[i]
  norm = Math.sqrt(norm)
  if (norm < MIN_EXTENT) return null
  const inv = 1 / norm
  for (let i = 0; i < DIMS; i++) vector[i] *= inv

  return { vector, weights, bbox: [x0, y0, x1, y1], visible }
}

/** Reflect a pose about the vertical axis, swapping left/right identities. */
export function mirrorKeypoints(kp: Keypoints): Keypoints {
  const out = new Float32Array(NUM_JOINTS * 3)
  for (let j = 0; j < NUM_JOINTS; j++) {
    const src = MIRROR_PERM[j]
    out[j * 3] = 1 - kp[src * 3]
    out[j * 3 + 1] = kp[src * 3 + 1]
    out[j * 3 + 2] = kp[src * 3 + 2]
  }
  return out
}

/** How many of the four shoulder/hip joints cleared the floor. */
export function torsoCount(kp: Keypoints, confFloor: number): number {
  let n = 0
  for (const j of [5, 6, 11, 12]) if (kp[j * 3 + 2] >= confFloor) n++
  return n
}

File: preprocessing/bodypose/encoding.py
"""The pose vector encoding. This module is the specification.

Everything downstream — the on-disk index, the browser matcher in
``live/src/encoding.ts`` — has to agree with what happens here, so keep the two
implementations in lockstep. ``EncodingParams`` is written into ``index.json``
and read back by the browser, so a mismatch is loud rather than silent.

The pipeline, per detected person:

  1. Take 17 COCO keypoints as ``(x, y, confidence)`` in image-normalised,
     top-left-origin coordinates. Every backend converts into this convention.
  2. Mark a joint visible when ``confidence >= conf_floor``.
  3. Bounding box over the visible joints only.
  4. Translate (bbox centre or corner) and scale (uniformly by the longer bbox
     side, or independently per axis) so the vector encodes limb configuration
     and nothing about where the figure sat in frame.
  5. Invisible joints collapse to the origin, which is neutral once weighted.
  6. Flatten to 34 floats and L2-normalise.

Cosine similarity between two of these is the recall metric; the per-joint
confidences ride along as weights for the re-rank stage.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, asdict
from typing import Sequence

ENCODING_VERSION = "coco17-v1"

#: Canonical joint order. Both backends map into this; the browser mirrors it.
COCO17 = (
    "nose",
    "left_eye",
    "right_eye",
    "left_ear",
    "right_ear",
    "left_shoulder",
    "right_shoulder",
    "left_elbow",
    "right_elbow",
    "left_wrist",
    "right_wrist",
    "left_hip",
    "right_hip",
    "left_knee",
    "right_knee",
    "left_ankle",
    "right_ankle",
)

NUM_JOINTS = len(COCO17)
DIMS = NUM_JOINTS * 2

JOINT_INDEX = {name: i for i, name in enumerate(COCO17)}

#: Index permutation that swaps left and right. Used for mirroring a pose.
MIRROR_PERM = tuple(
    JOINT_INDEX[
        name.replace("left_", "right_") if name.startswith("left_")
        else name.replace("right_", "left_") if name.startswith("right_")
        else name
    ]
    for name in COCO17
)

TORSO = tuple(JOINT_INDEX[n] for n in ("left_shoulder", "right_shoulder", "left_hip", "right_hip"))

# A degenerate bounding box (a figure seen edge-on, or a detection that
# collapsed) would otherwise divide by ~zero and produce a vector of garbage
# that happens to have unit norm.
MIN_EXTENT = 1e-4


@dataclass(frozen=True)
class EncodingParams:
    """Knobs that must match between the indexer and the browser."""

    version: str = ENCODING_VERSION
    #: "center" puts the bbox centre at the origin; "corner" uses the top-left,
    #: which is what Move Mirror did.
    origin: str = "center"
    #: "uniform" divides both axes by the longer bbox side, preserving aspect
    #: ratio. "independent" divides each axis by its own extent, the Move Mirror
    #: behaviour — it makes every pose fill a unit square, which throws away the
    #: difference between a wide stance and a narrow one.
    scale: str = "uniform"
    #: Below this confidence a joint is treated as absent.
    conf_floor: float = 0.1

    def validate(self) -> None:
        if self.origin not in ("center", "corner"):
            raise ValueError(f"origin must be 'center' or 'corner', got {self.origin!r}")
        if self.scale not in ("uniform", "independent"):
            raise ValueError(f"scale must be 'uniform' or 'independent', got {self.scale!r}")
        if not 0.0 <= self.conf_floor < 1.0:
            raise ValueError(f"conf_floor must be in [0, 1), got {self.conf_floor}")

    def to_json(self) -> dict:
        d = asdict(self)
        d["keypoints"] = list(COCO17)
        d["dims"] = DIMS
        d["y_axis"] = "top-left"
        d["min_extent"] = MIN_EXTENT
        return d


@dataclass
class EncodedPose:
    vector: list[float]        # 34 floats, unit L2 norm
    weights: list[float]       # 17 confidences, clamped to [0, 1]
    bbox: tuple[float, float, float, float]  # x0, y0, x1, y1 in image coords
    visible: int               # number of joints above conf_floor


def encode(
    keypoints: Sequence[Sequence[float]],
    params: EncodingParams = EncodingParams(),
) -> EncodedPose | None:
    """Turn 17 ``(x, y, confidence)`` triples into a comparable unit vector.

    Returns ``None`` when the pose is too degenerate to encode — no visible
    joints, or a bounding box with no extent in either axis.
    """
    if len(keypoints) != NUM_JOINTS:
        raise ValueError(f"expected {NUM_JOINTS} keypoints, got {len(keypoints)}")

    xs: list[float] = []
    ys: list[float] = []
    cs: list[float] = []
    for x, y, c in keypoints:
        xs.append(float(x))
        ys.append(float(y))
        cs.append(min(1.0, max(0.0, float(c))))

    vis = [c >= params.conf_floor for c in cs]
    n_vis = sum(vis)
    if n_vis == 0:
        return None

    vx = [xs[i] for i in range(NUM_JOINTS) if vis[i]]
    vy = [ys[i] for i in range(NUM_JOINTS) if vis[i]]
    x0, x1 = min(vx), max(vx)
    y0, y1 = min(vy), max(vy)
    w = x1 - x0
    h = y1 - y0
    if w < MIN_EXTENT and h < MIN_EXTENT:
        return None

    if params.origin == "center":
        ox, oy = (x0 + x1) * 0.5, (y0 + y1) * 0.5
    else:
        ox, oy = x0, y0

    if params.scale == "uniform":
        sx = sy = max(w, h, MIN_EXTENT)
    else:
        sx, sy = max(w, MIN_EXTENT), max(h, MIN_EXTENT)

    vec = [0.0] * DIMS
    for i in range(NUM_JOINTS):
        if vis[i]:
            vec[i * 2] = (xs[i] - ox) / sx
            vec[i * 2 + 1] = (ys[i] - oy) / sy
        # else: leave at the origin, neutral once the weight is applied

    norm = math.sqrt(sum(v * v for v in vec))
    if norm < MIN_EXTENT:
        return None
    inv = 1.0 / norm
    vec = [v * inv for v in vec]

    weights = [cs[i] if vis[i] else 0.0 for i in range(NUM_JOINTS)]
    return EncodedPose(vector=vec, weights=weights, bbox=(x0, y0, x1, y1), visible=n_vis)


def mirror_keypoints(keypoints: Sequence[Sequence[float]]) -> list[list[float]]:
    """Reflect a pose about the vertical axis of the image.

    Flips x within the unit image box and swaps the left/right joint identities,
    so a figure facing one way becomes the same figure facing the other.
    """
    return [
        [1.0 - float(keypoints[src][0]), float(keypoints[src][1]), float(keypoints[src][2])]
        for src in MIRROR_PERM
    ]


def torso_visible(keypoints: Sequence[Sequence[float]], conf_floor: float, need: int = 3) -> bool:
    """Whether enough of the shoulder/hip quad survived to trust the bbox."""
    return sum(1 for i in TORSO if keypoints[i][2] >= conf_floor) >= need


def cosine(a: Sequence[float], b: Sequence[float]) -> float:
    return sum(x * y for x, y in zip(a, b))


def weighted_distance(
    va: Sequence[float], wa: Sequence[float],
    vb: Sequence[float], wb: Sequence[float],
) -> float:
    """Confidence-weighted L2 between two encoded poses.

    Each joint contributes in proportion to the product of both detectors'
    confidence in it, so an ankle one side was guessing at cannot drag the
    match around. Normalised by total weight, so it stays comparable across
    poses with different numbers of visible joints.
    """
    total = 0.0
    acc = 0.0
    for j in range(NUM_JOINTS):
        omega = wa[j] * wb[j]
        if omega <= 0.0:
            continue
        dx = va[j * 2] - vb[j * 2]
        dy = va[j * 2 + 1] - vb[j * 2 + 1]
        acc += omega * (dx * dx + dy * dy)
        total += omega
    if total <= 0.0:
        return float("inf")
    return math.sqrt(acc / total)
